### PR TITLE
⚠️ Make machine.spec.version required

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
machine.spec.version is a *string, but this field is actually required and the controller fails if it is missing.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/4996

### CODE REVIEW

**Comment by sbueringer** [comment]
lgtm apart from @ykakarap's nit

**Comment by fabriziopandini** [comment]
Comment addressed + commit squashed

File: api/v1alpha4/conversion_test.go [hunk 1]
```diff
@@ -58,6 +60,21 @@ func TestFuzzyConversion(t *testing.T) {
...
+	}
+}
+
+func MachineSpecFuzzer(in *MachineSpec, c fuzz.Continue) {
+	c.FuzzNoCustom(in)
+
+	// Version field has been converted from *string to string in v1beta1,
+	// so we're forcing valid string values to avoid round trip errors.
+	versionString := c.RandString()
+	in.Version = &versionString
```
Comment (vincepri) [inline api/v1alpha4/conversion_test.go:75]: It seems here the problem only appears when in.Version has been generated to be a `nil` value. Could we only force a string here `if in.Version == nil {` ?

Comment (fabriziopandini) [reply api/v1alpha4/conversion_test.go]: Done!

**Comment by enxebre** [comment]
I think is positive for this to be optional and to be an overriding choice for bootstrap providers as today. This wouldn't fit any existing bootstrapping workflows relying on DataSecretName where the version might be dictated by something else or different to the kubeadm referenced here.

**Comment by vincepri** [comment]
@enxebre Wouldn't the version be at least informational in that case?

**Comment by enxebre** [comment]
It could be I guess but it becomes problematic for some uses cases I can think of, e.g a change to the version is a change to template spec and so it would unavoidable trigger a rolling upgrade which might or might no me desirable for a given bootstrapper or might be meaningless without an in hand change to the user data which might be the source of truth. Another use case might be using user data for in place upgrades which I anticipate it will be a hard requirement for baremetal envs where recreation type upgrades are not always assumable.
I think I see no harm in keeping this optional whereas making it required would introduce an unnecessary rigidness.

**Comment by vincepri** [comment]
How would you reconcile these requirements with what @chymy pointed out in the issue?

**Comment by enxebre** [comment]
>How would you reconcile these requirements with what @chymy pointed out in the issue?

I see that as bootstrapper choice while honouring the [current assumption](https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/machine_types.go#L70-L73)
If the field is set the bootstrap provider should use it, otherwise use its own. If none is set then it should report a failure appropriately i.e logging/conditions.
Then is up to higher level tooling to set the right input e.g clusterClass.

**Comment by vincepri** [comment]
@enxebre The comment on the field is outdated, the field today is not only used by bootstrap providers, but also all major infrastructure providers today use it to pick a VM image, which is what the issue is about.

cc @yastij @randomvariable @CecileRobertMichon

**Comment by enxebre** [comment]
>@enxebre The comment on the field is outdated, the field today is not only used by bootstrap providers, but also all major infrastructure providers today use it to pick a VM image, which is what the issue is about.

Oh right I see, like in aws if you don't set an explicit ami, thanks for clarifying. We should update https://github.com/kubernetes-sigs/cluster-api/issues/4996 to reflect that since it's only pointing to some kubeadmconfig_controller.go logs atm.

I'd reiterate my thinking above about keeping current flexibility and decoupling as a choice. It provides great value in the context of Immutable operating systems combined with ignition and in baremetal envs.
As we evolve we can always make this field required if we choose to but as of today I think having it optional provides good value by easing and enabling multiple use cases that enrich the project and feed us with valuable feedback beyond the reference implementation to keep evolving. Happy to hear other's/providers thoughts.

**Comment by fabriziopandini** [comment]
As per Sept 29th Office hours discussion

**Pull request closed without merging**